### PR TITLE
Add a string copy function

### PR DESCRIPTION
**Pull request by david-koenig**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/string.c [hunk 1]
```diff
@@ -19,18 +19,29 @@ struct aws_string *aws_string_new_from_c_str(struct aws_allocator *allocator, co
...
+    *(struct aws_allocator **)(&str->allocator) = allocator;
+    *(size_t *)(&str->len) = len;
+    memcpy((void *)str->bytes, bytes, len);
+    *(uint8_t *)&str->bytes[len] = '\0';
 
-    return hdr;
+    return str;
+}
+
+struct aws_string *aws_string_copy(struct aws_allocator *allocator, const struct aws_string *str) {
```
Comment (bdonlan) [inline source/string.c]: Better to call this _clone or _dup? _copy (without an _init or _new) implies to me that it's copying to an existing destination

File: source/string.c [hunk 2]
```diff
@@ -19,18 +19,29 @@ struct aws_string *aws_string_new_from_c_str(struct aws_allocator *allocator, co
...
+    size_t size = sizeof(struct aws_string) + str->len + 1;
```
Comment (bdonlan) [inline source/string.c]: Delegate to aws_string_new_from_array.

File: source/string.c (hunk 1 above)
Comment (ColdenCullen) [reply source/string.c]: I like _dup, it matches what C's `strdup`.

Comment (david-koenig) [reply source/string.c]: We settled on ``aws_string_new_from_string`` based on conversation with the team.

File: source/string.c (hunk 2 above)
Comment (david-koenig) [reply source/string.c]: Done.

**Pull request merged**